File: 1579-remove-max-number-of-edges-to-keep-graph-fully-traversable/1579-remove-max-number-of-edges-to-keep-graph-fully-traversable.cpp
class DSU {
    vector<int> parent;
    int remaining;
public:
    DSU(int n) {
        parent.resize(n+1);
        for (int i = 1; i <= n; ++i) {
            parent[i] = i;
        }
        remaining = n;
    }
    
    int find(int x) {
        if(parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }
    
    bool unionfind(int a, int b) {
        int pa = find(a);
        int pb = find(b);
        if(pa==pb) return false;
        parent[pb] = pa;
        remaining--;
        return true;
    }
    
    int getRemaining() {
        return remaining;
    }
};

class Solution {
public:
    int maxNumEdgesToRemove(int n, vector<vector<int>>& edges) {
        // sort(edges.begin(), edges.end(), [](vector<int>&e1, vector<int>&e2) {return e1[0]>e2[0];});
        
        DSU alice(n), bob(n);
        
        int ans = 0;
        for (int i = 3; i>=1;i--) {
            for (auto edge : edges) {
                if (edge[0] != i) continue;
                if (edge[0] == 1) {
                    if(!alice.unionfind(edge[1], edge[2])) ans++;
                } else if (edge[0] == 2) {
                    if(!bob.unionfind(edge[1], edge[2])) ans++;
                } else {
                    if ((!alice.unionfind(edge[1], edge[2])) || (!bob.unionfind(edge[1], edge[2]))) ans++;
                }
            }
        }
        
        if (alice.getRemaining() != 1 || bob.getRemaining() != 1) return -1;
        return ans;
    }
};

File: 1579-remove-max-number-of-edges-to-keep-graph-fully-traversable/1579-remove-max-number-of-edges-to-keep-graph-fully-traversable.java
class DSU {
    int n, count;
    int[] parent;
    
    public DSU(int num) {
        n = num+1;
        count = num;
        parent = new int[n];
        
        for (int i = 0; i < n; ++i) {
            parent[i] = i;
        }
    }
    
    public int find(int a) {
        if (parent[a] != a) {
            parent[a] = find(parent[a]);
        }
        return parent[a];
    }
    
    public boolean unionfind(int a, int b) {
        int pa = find(a);
        int pb = find(b);
        
        if (pa == pb) {
            return false;
        }
        count--;
        parent[pb] = pa;
        return true;
    }
    
    public boolean allCovered() {
        return (count==1);
    }
}

class Solution {
    public int maxNumEdgesToRemove(int n, int[][] edges) {
        DSU dsu1 = new DSU(n);
        DSU dsu2 = new DSU(n);
        
        int answer = 0;
        for (int[] edge : edges) {
            if (edge[0] == 3) {
                if (!dsu1.unionfind(edge[1], edge[2])) {
                    answer++;
                }
                
                dsu2.unionfind(edge[1], edge[2]);
            }
        }
        
        
        for (int[] edge : edges) {
            if (edge[0] == 1) {
                if (!dsu1.unionfind(edge[1], edge[2])) {
                    answer++;
                }
            } else if(edge[0] == 2) {
                if (!dsu2.unionfind(edge[1], edge[2])) {
                    answer++;
                }
            }
        }
        
        return (dsu1.allCovered() && dsu2.allCovered())?answer:-1;
    }
}